lib/BridgedRoom.js

```diff
@@ -0,0 +1,123 @@
+
+      events.on('chatMessages', (message) => {
+        if (!message.model.fromUser || message.model.fromUser.id == gitterUserId) {
+          // Ignore a reflection of my own messages
+          return;
+        }
+
+        if (message.operation !== 'create' && message.operation !== 'update') {
+          return;
+        }
+
+        this.onGitterMessage(message);
+      });
+
+      this._gitterRealtime.subscribe("/v1/rooms/" + room.id, (message) => {
+        this._bridge.getGitterUserById(message.userId).then((user) => {
+          if (user)
+            user.setRoomPresence(room.id, message.status == 'in');
+        });
+      });
+    });
+  });
+};
+
+BridgedRoom.prototype.stopAndLeave = function() {
+  this._gitterRoom.streaming().disconnect();
+
+  return this._gitterRoom.removeUser(this._gitterUserId);
+};
+
+function quotemeta(s) { return s.replace(/\W/g, '\\$&'); }
+
+BridgedRoom.prototype.onGitterMessage = function(message) {
+  var fromUser = message.model.fromUser;
+
+  console.log('gitter->' + this.gitterRoomName() + ' from ' + fromUser.username + ':', message.model.text)
+
+  this._bridge.mapGitterUser(fromUser).then((user) => {
+    return user.update(fromUser)
+      .catch((e) => {
+        console.log("Updating user failed:", e);
+        // ignore the failure and continue anyway
+      }).then(() => {
+        return user;
+      });
+  }).then((user) => {
+    var matrixMessage = {
+      msgtype: "m.text",
+      body: message.model.text,
+    };
+
+    // Pull out the HTML part of the body if it's not just plain text
+    if (message.model.html != message.model.text) {
+      matrixMessage["format"] = "org.matrix.custom.html";
+      matrixMessage["formatted_body"] = message.model.html;
+    }
+
+    if (message.model.status) {
+      matrixMessage["msgtype"] = "m.emote";
+
+      // Strip the leading @username mention from the body text
+      var userNameQuoted = quotemeta(fromUser.username);
+
+      matrixMessage["body"] =
+        matrixMessage["body"].replace(new RegExp("^@" + userNameQuoted + " "), "");
+
+      // TODO(paul): HTML is harder. Applying regexp mangling to an HTML string.
+      //   This is terrible. kegan - please help ;)
+      matrixMessage["formatted_body"] =
+        matrixMessage["formatted_body"].replace(new RegExp("^<span [^>]+>@" + userNameQuoted + "</span> "), "");
+    }
+
+    // TODO(paul): consider if we want to annotate somehow if message.operation was
+    //   'update'
+
+    return user.getIntent().sendMessage(this.matrixRoomId(), matrixMessage);
+  });
+};
+
+BridgedRoom.prototype.onMatrixMessage = function(message) {
+  var from = this._bridge.mangleName(message.user_id);
+
+  // gitter supports Markdown. We'll use that to apply a little formatting
+  // to make understanding the text a little easier
+  if (message.content.msgtype == 'm.emote') {
+    var text = '*' + (from + ' ' + message.content.body).replace(/\*/g, '\\*') + '*';
```

Comment:
I'm going to say this here but this can be applied basically everywhere: please document what your intent is when mangling strings. I do not know if what you're doing here is what you *intend* to do, unless I can see an example input/output e.g:

```javascript
// Surround emotes in *s and add their name, escaping any *s in the message e.g:
// "thinks for * * * a bit"  =>  "*Kegan thinks for \* \* \* a bit *"
```